core/src/main/java/org/apache/calcite/sql/fun/SqlJsonValueFunction.java

```diff
@@ -54,20 +53,33 @@ public class SqlJsonValueFunction extends SqlFunction {
 
   public SqlJsonValueFunction(String name) {
     super(name, SqlKind.OTHER_FUNCTION,
-        ReturnTypes.cascade(
-            opBinding -> explicitTypeSpec(opBinding).orElse(getDefaultType(opBinding)),
-            SqlTypeTransforms.FORCE_NULLABLE),
+        opBinding -> explicitTypeSpec(opBinding)
+            .map(relDataType -> deriveExplicitType(opBinding, relDataType))
+            .orElseGet(() -> getDefaultType(opBinding)),
         null,
         OperandTypes.family(
             ImmutableList.of(SqlTypeFamily.ANY, SqlTypeFamily.CHARACTER),
             ordinal -> ordinal > 1),
         SqlFunctionCategory.SYSTEM);
   }
 
+  private static RelDataType deriveExplicitType(SqlOperatorBinding opBinding, RelDataType type) {
+    if (SqlTypeName.ARRAY == type.getSqlTypeName()) {
+      RelDataType elementType = Objects.requireNonNull(type.getComponentType());
+      RelDataType nullableElementType = deriveExplicitType(opBinding, elementType);
+      return SqlTypeUtil.createArrayType(
+          opBinding.getTypeFactory(),
+          nullableElementType,
+          true);
+    }
+    return opBinding.getTypeFactory().createTypeWithNullability(type, true);
+  }
+
   /** Returns VARCHAR(2000) as default. */
   private static RelDataType getDefaultType(SqlOperatorBinding opBinding) {
     final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
-    return typeFactory.createSqlType(SqlTypeName.VARCHAR, 2000);
+    final RelDataType baseType = typeFactory.createSqlType(SqlTypeName.VARCHAR, 2000);
```

Comment:
why 2000? This will truncate results to 2000 characters. why not just VARCHAR?

Reply:
i don't have the context for why this was originally chosen, i was just preserving existing behavior

Reply:
I have fixed a couple of bugs in the past related to that: https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-5813. This very much looks like a bug to me. It means that if you get a result longer than 2000 characters it will be truncated. And that's not what you want. Whether you want to fix it in this PR or another one, that's a different issue.

Reply:
yeah, happy to change it in this PR, it seems like a strange default to me

Reply:
looking closer, it seems like a lot of the JSON functions have varchar 2000 as the default. I'm kind of thinking should split that change into a separate PR to change them all at once instead of just fixing one (or all of them) as part of this PR

Reply:
sounds reasonable